### PR TITLE
[SDK-3733] Make realtime transports tree-shakable

### LINKED ISSUE #1394: Create tree-shakable transport modules
- depends on #1374 and #1393 
- create new  `WebSocketTransport`, `XHRPolling` and `XHRStreaming` classes
- These transports may be used by a `BaseClient` if and only if they are provided as constructor arguments

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
**Note: This PR is based on top of #1429; please review that one first.**

This exposes tree-shakable `WebSocketTransport`, `XHRPolling`, and `XHRStreaming` modules.

See commit messages for more details.

Resolves #1394.

### CODE REVIEW

**Review by owenpearson — CHANGES_REQUESTED** [review]
implementation looks great, one comment about the modules api though:

File: src/common/lib/client/baserealtime.ts [hunk 1]
```diff
@@ -11,24 +11,45 @@ import ClientOptions from '../../types/ClientOptions';
...
     super(options, modules);
     Logger.logAction(Logger.LOG_MINOR, 'Realtime()', '');
+    this._additionalTransportImplementations = BaseRealtime.transportImplementationsFromModules(modules);
     this._RealtimePresence = modules.RealtimePresence ?? null;
     this.connection = new Connection(this, this.options);
     this._channels = new Channels(this);
     if (options.autoConnect !== false) this.connect();
   }
 
+  private static transportImplementationsFromModules(modules: ModulesMap) {
```
[the remark this answers is not in the document]
Comment (lawrence-forooghian) [reply src/common/lib/client/baserealtime.ts:37]: I'm afraid I don't know what you mean here — what does it mean to "pass an invalid transport" and what is it that we're currently ignoring?

[the remark this answers is not in the document]
Comment (lawrence-forooghian) [reply src/common/lib/client/baserealtime.ts:37]: Perhaps we could merge it as it is for now? I imagine we'll want to give the API a once-over before release anyway.

[the remark this answers is not in the document]
Comment (owenpearson) [reply src/common/lib/client/baserealtime.ts:37]: yes, sgtm 👍

**Review by owenpearson — APPROVED** [review]

**Pull request merged**